micro_benchmarking_pytorch.py

```diff
@@ -215,13 +215,39 @@ def run_benchmarking(local_rank, ngpus, net, batch_size, iterations, prof_step,
 
     ## benchmark.
     print ("INFO: running the benchmark..")
-    tm = time.time()
-    for i in range(iterations):
-        if i == prof_step:
-            forwardbackward(inp, optimizer, network, target, amp_opt_level, i)
-        else:
-            forwardbackward(inp, optimizer, network, target, amp_opt_level)
-    torch.cuda.synchronize()
+    if args.kineto:
+        from torch.profiler import schedule, profile, ProfilerActivity, record_function
+        profiler_schedule = schedule(
+            skip_first = 0,
+            wait = 1,
+            warmup = 2,
+            active = 2,
+            repeat = 1,
+        )
+
+        def trace_ready_callback(prof):
+            print("----------- Trace Ready -----------")
+            prof.export_chrome_trace(f"trace{prof.step_num}.json")
+
+        tm = time.time()
+        with profile(
+            activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
+            schedule=profiler_schedule,
+            on_trace_ready=trace_ready_callback) as prof:
+            for i in range(iterations):
+                with record_function(f"iteration {i}"):
+                    forwardbackward(inp, optimizer, network, target, amp_opt_level)
+                prof.step()
+            torch.cuda.synchronize()
+            print(prof.key_averages().table(sort_by="cuda_time_total"))
+    else:
+        tm = time.time()
+        with torch.autograd.profiler.emit_nvtx(enabled=args.autograd_profiler):
+            for i in range(iterations):
+                if i == flops_prof_step:
+                    forwardbackward(inp, optimizer, network, target, amp_opt_level, i)
+                else:
+                    forwardbackward(inp, optimizer, network, target, amp_opt_level)
```

Comment:
@jeffdaily @mwootton We missed a bug here: I accidentally removed the `torch.cuda.synchronize()` call in the else (autograd_profiler) clause, causing it to undercount the time (eg. https://ontrack-internal.amd.com/browse/SWDEV-341071?focusedCommentId=9009931&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-9009931). I fixed the bug here: https://github.com/ROCmSoftwarePlatform/pytorch-micro-benchmarking/commit/7542f973c737dccd5e46d0a9cce557971dfcc2f3